src/authorization/access-checkers/AgentGroupAccessChecker.ts

```diff
@@ -0,0 +1,77 @@
+
+import type { ResourceIdentifier } from '../../ldp/representation/ResourceIdentifier';
+import type { RepresentationConverter } from '../../storage/conversion/RepresentationConverter';
+import type { ExpiringStorage } from '../../storage/keyvalue/ExpiringStorage';
+import { fetchDataset } from '../../util/FetchUtil';
+import { promiseSome } from '../../util/PromiseUtil';
+import { readableToQuads } from '../../util/StreamUtil';
+import { ACL, VCARD } from '../../util/Vocabularies';
+import type { AccessCheckerArgs } from './AccessChecker';
+import { AccessChecker } from './AccessChecker';
+
+/**
+ * Checks if the given WebID belongs to a group that has access.
+ *
+ * Fetched results will be stored in an ExpiringStorage.
+ *
+ * Requires a storage that can store JS objects.
+ * `expiration` parameter is how long entries in the cache should be stored in minutes, defaults to 60.
+ */
+export class AgentGroupAccessChecker extends AccessChecker {
+  private readonly converter: RepresentationConverter;
+  private readonly cache: ExpiringStorage<string, Promise<Store>>;
+  private readonly expiration: number;
+
+  public constructor(converter: RepresentationConverter, cache: ExpiringStorage<string, Promise<Store>>,
+    expiration = 60) {
+    super();
+    this.converter = converter;
+    this.cache = cache;
+    // Convert minutes to milliseconds
+    this.expiration = expiration * 60 * 1000;
+  }
+
+  public async handle({ acl, rule, credentials }: AccessCheckerArgs): Promise<boolean> {
+    if (typeof credentials.webId === 'string') {
+      const { webId } = credentials;
+      const groups = acl.getObjects(rule, ACL.terms.agentGroup, null);
+
+      return await promiseSome(groups.map(async(group: Term): Promise<boolean> =>
+        this.isMemberOfGroup(webId, group)));
+    }
+    return false;
+  }
+
+  /**
+   * Checks if the given agent is member of a given vCard group.
+   * @param webId - WebID of the agent that needs access.
+   * @param group - URL of the vCard group that needs to be checked.
+   *
+   * @returns If the agent is member of the given vCard group.
+   */
+  private async isMemberOfGroup(webId: string, group: Term): Promise<boolean> {
+    const groupDocument: ResourceIdentifier = { path: /^[^#]*/u.exec(group.value)![0] };
+
+    // Fetch the required vCard group file
+    const quads = await this.fetchCachedQuads(groupDocument.path);
+    return quads.countQuads(group, VCARD.terms.hasMember, webId, null) !== 0;
+  }
+
+  /**
+   * Fetches quads from the given URL.
+   * Will cache the values for later re-use.
+   */
+  private async fetchCachedQuads(url: string): Promise<Store> {
+    let result = await this.cache.get(url);
+    if (!result) {
+      const prom = (async(): Promise<Store> => {
+        const representation = await fetchDataset(url, this.converter);
+        return readableToQuads(representation.data);
+      })();
+      await this.cache.set(url, prom, this.expiration);
+      result = await prom;
+    }
```

Comment:
```suggestion
    let result = this.cache.get(url);
    if (!result) {
      result = (async(): Promise<Store> => {
        const representation = await fetchDataset(url, this.converter);
        return readableToQuads(representation.data);
      })();
      await this.cache.set(url, result, this.expiration);
    }
```

Reply:
Typings don't allow for the reassignment since the thing you get out of the store is `Promise<Promise<Store> | undefined>`, or `Store | undefined` if you await, but the return type of that function is `Promise<Store>`